src/telemetry/storage_usage.rs

```diff
@@ -0,0 +1,79 @@
+/*
+ * This file is part of Edgehog.
+ *
+ * Copyright 2022 SECO Mind Srl
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+use crate::error::DeviceManagerError;
+use astarte_sdk::types::AstarteType;
+use serde::Serialize;
+use std::collections::HashMap;
+use sysinfo::{DiskExt, System, SystemExt};
+
+#[derive(Debug, Serialize)]
+#[serde(rename_all = "camelCase")]
+pub struct DiskUsage {
+    pub disk_name: String,
+    pub total_bytes: i64,
+    pub free_bytes: i64,
+}
+
+/// get available partition disks on the system
+fn get_disks() -> Result<Vec<DiskUsage>, DeviceManagerError> {
+    let mut ret: Vec<DiskUsage> = Vec::new();
+    let mut sys = System::new_all();
+    sys.refresh_disks();
+
+    for disk in sys.disks() {
+        let mut map: HashMap<String, AstarteType> = HashMap::new();
+        let disk_name = disk.name().to_str().unwrap().replace("/dev", "");
+        map.insert(
+            format!("{}/totalBytes", disk_name),
+            (disk.total_space() as i64).into(),
+        );
+        map.insert(
+            format!("{}/freeBytes", disk_name),
+            (disk.available_space() as i64).into(),
+        );
+        ret.push(DiskUsage {
+            disk_name: disk_name.to_owned() as String,
+            total_bytes: disk.total_space() as i64,
+            free_bytes: disk.available_space() as i64,
+        });
+    }
+    Ok(ret)
+}
+
+/// get structured data for `io.edgehog.devicemanager.StorageUsage` interface
+pub fn get_storage_usage() -> Result<HashMap<String, AstarteType>, DeviceManagerError> {
```

Comment:
You can use a struct instead of an HashMap, since StorageUsage is an object, similar to get_system_status

Reply:
Struct is used in getDisks already and here Hashmap is required to format as per astarte interface

Reply:
please look at get_system_status

Reply:
I looked already into get_system_status. It is not similar. I cannot use a struct here because I need to construct dynamic field names at runtime inside the struct.